### PR TITLE
add --pre-deploy-script arg to robotpy-installer

### PR DESCRIPTION
**Pull request by otherjon**
Hi robotpy team,

I was wondering how you folks might feel about allowing a pre-release script to be specified on the command line when deploying code to a RoboRIO.

A pre-deploy script can serve the same purpose as Gradle tasks, e.g. to determine the Git commit hash and branch and write them to files to be deployed to the RoboRIO.  (See [this WPILib page](https://docs.wpilib.org/en/stable/docs/software/advanced-gradlerio/deploy-git-data.html) for the Java equivalent.)

In this draft implementation, pre-deploy scripts must be written in Python (this allows equivalent invocation on Windows or Linux), and will receive a command-line argument of --robot_path <path/to/dir/of/robot.py>

Sample predeploy script: [available here](https://github.com/otherjon/Project212-swervetest/blob/main/predeploy.py)

I see that you're in the process of moving deploy functionality from pyfrc to robot-installer, so I'm submitting my pull request as a commit on top of the robotpy-installer/add-deploy branch, but let me know if there's a better strategy.

### CODE REVIEW

**Comment by auscompgeek** [comment]
> e.g. to determine the Git commit hash and branch and write them to files to be deployed to the RoboRIO.

- We already gather git metadata to stamp the deploy, which gets printed at robot start. See https://robotpy.readthedocs.io/en/stable/dev/deploy.html#deploy-artifacts

- Running other arbitrary code at deploy time is also already possible, as your robot code is the entry point for deploying. This is how I stamped deploys with git metadata before we baked it into the deploy process as outlined above.

Do you have a use-case which isn't well suited by the above? I'm not sure this complexity makes sense here.

**Comment by otherjon** [comment]
Thanks for the pointer to the deploy artifact information.  I hadn't looked at the developer documentation earlier because I hadn't considered myself a robotpy developer until now.  Since the Git info was my primary motivation, I'll cancel the pull request.

While obviously I run "robot.py deploy" in order to deploy, I'm at a loss to explain how I'd add code to my robot.py to execute a pre-release script.  Do you have a pointer to info about that, or the name a method I should override?

File: robotpy_installer/deploy.py [hunk 1]
```diff
@@ -169,6 +175,28 @@ def run(self, options, robot_class, **static_options):
...
+            try:
+                # The pre-deploy script will be executed with the user's
+                # permissions.  We assume this is not a security issue,
+                # since the user is explicitly choosing on the command line
+                # to run the script.  If a pre-deploy script could be run
+                # automatically (based on its name, e.g.), there might be
+                # security concerns, especially when checking out an
+                # external repo to test someone else's code.
+                #
+                args = ['python3', predeploy, '--robot_path', robot_path]
```
Comment (virtuald) [inline robotpy_installer/deploy.py:194]: If you were to do this, always use `sys.executable` not `python3` directly to ensure they're ran with the same interpreter.

File: robotpy_installer/deploy.py [hunk 2]
```diff
@@ -169,6 +175,28 @@ def run(self, options, robot_class, **static_options):
...
+                subprocess.check_output(args)
```
Comment (virtuald) [inline robotpy_installer/deploy.py:195]: You aren't using the output of the script, so you'd want to use `subprocess.run(... check=True)` instead.

File: robotpy_installer/deploy.py [hunk 3]
```diff
@@ -169,6 +175,28 @@ def run(self, options, robot_class, **static_options):
...
+            except Exception as e:
```
Comment (virtuald) [inline robotpy_installer/deploy.py:196]: This would want to catch the more specific `CalledProcessError` instead of any exception.

**Review by virtuald — COMMENTED** [review]
I think the idea is fine, but I don't really like this implementation. In particular, if the script is so important to run, why does the user have to specify it on the command line each time? What if they forget? It just feels like you should make the script as part of whatever custom bash script you run wrapping this, or run something in-process.

I'd rather either:

* explicitly named file loaded automatically (like how physics.py works... but that doesn't really feel like a good fit her)
* Some config file specifies which file to run
* Provide a mechanism in `wpilib.run` to call a function in your program with relevant information when alll/specific commands are ran

**Comment by virtuald** [comment]
It's also documented in the user docs at https://robotpy.readthedocs.io/projects/robotpy/en/latest/wpilib.deployinfo.html, but... it wasn't documented in prior years even though it was there if you looked for it.

Strictly speaking, you could put whatever arbitrary code you want in the `if __name__ == "__main__":` block of your code, including something that checks sys.argv to see if deploy is occurring.

**Pull request closed without merging**